File: ios/SiriApp/ViewModels/MainViewModel.swift
//
//  MainViewModel.swift
//  SiriApp
//
//  Main pipeline orchestrator: ASR → LLM → TTS → Audio playback.
//  Ported from Android: MainViewModel.kt
//

import Foundation
import Combine
import os.log

@MainActor
class MainViewModel: ObservableObject {
    @Published var state = AppState()
    @Published var messages: [ChatMessage] = []
    @Published var assistantReply: String = ""

    private let audioRecorder = AudioRecorder()
    private let audioPlayer = AudioPlayer()
    private let configRepo = ConfigRepository()
    private let documentsDir: URL

    private lazy var asrEngine: SherpaAsrEngine = {
        SherpaAsrEngine(documentsDir: documentsDir)
    }()
    private lazy var ttsEngine: SherpaTtsEngine = {
        SherpaTtsEngine(documentsDir: documentsDir)
    }()
    private lazy var llmClient = LlmClient(configRepository: configRepo)
    private(set) lazy var chatSession = ChatSession(llmClient: llmClient)

    private var recordingCancellable: AnyCancellable?
    private var streamingCancellable: AnyCancellable?
    private var streamingContinuation: CheckedContinuation<Void, Never>?
    private var speakingTask: Task<Void, Never>?
    private var recognitionTask: Task<Void, Never>?
    private var engineCleanup: (() -> Void)?

    init() {
        documentsDir = FileManager.default.urls(
            for: .documentDirectory, in: .userDomainMask
        )[0]

        // Load TTS preference
        state.ttsEnabled = UserDefaults.standard.object(forKey: "tts_enabled") as? Bool ?? false

        // Observe chat messages
        chatSession.$messages
            .receive(on: DispatchQueue.main)
            .assign(to: &$messages)

        // Audio interruption handling
        AudioSessionManager.startObservingInterruptions()
        AudioSessionManager.onInterruptionBegan = { [weak self] in
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                os_log(.info, "MainVM: handling audio interruption began")
                self.cancelListening()
            }
        }
    }

    // MARK: - Initialization

    func initializeEngines() {
        state.voiceState = .loading("模型加载中…")

        Task.detached(priority: .userInitiated) { [weak self] in
            guard let self = self else { return }

            let asrReady = await MainActor.run { self.asrEngine.initialize() }
            let ttsReady = await MainActor.run { self.ttsEngine.initialize() }

            await MainActor.run {
                self.state.enginesReady = asrReady && ttsReady
                self.state.voiceState = (asrReady && ttsReady)
                    ? .idle
                    : .error("模型加载失败，请检查模型文件")
            }

            // Store cleanup closure captured on main actor for deinit
            await MainActor.run { [weak self] in
                self?.engineCleanup = { [asr = self?.asrEngine, tts = self?.ttsEngine] in
                    asr?.destroy()
                    tts?.destroy()
                }
            }
        }
    }

    // MARK: - Config

    func checkConfig() -> Bool {
        let hasConfig = configRepo.hasConfig
        state.hasConfig = hasConfig
        if hasConfig, case .error(let msg) = state.voiceState, msg.contains("配置 API") {
            state.voiceState = .idle
        }
        return hasConfig
    }

    func toggleTts(_ enable: Bool) {
        state.ttsEnabled = enable
        UserDefaults.standard.set(enable, forKey: "tts_enabled")
        if !enable {
            stopSpeaking()
        }
    }

    // MARK: - Cancel all tasks

    private func cancelAllTasks() {
        os_log(.info, "MainVM: cancelAllTasks — aborting all in-progress work")

        // Coroutine tasks
        recordingCancellable?.cancel()
        recordingCancellable = nil
        streamingCancellable?.cancel()
        streamingCancellable = nil
        streamingContinuation?.resume()
        streamingContinuation = nil
        speakingTask?.cancel()
        speakingTask = nil
        recognitionTask?.cancel()
        recognitionTask = nil

        // Audio I/O
        audioRecorder.stop()
        audioPlayer.stop()

        // ASR engine buffer — discard stale samples from cancelled sessions
        asrEngine.reset()
    }

    // MARK: - Recording

    func startListening() {
        guard state.enginesReady else {
            state.voiceState = .error("模型未就绪，请先上传模型文件")
            return
        }

        // Cancel everything in flight — ASR, LLM, TTS, etc.
        cancelAllTasks()

        AudioSessionManager.configure()

        os_log(.info, "MainVM: start listening")
        state.voiceState = .listening
        state.partialAsrText = ""
        state.finalAsrText = ""

        recordingCancellable = audioRecorder.startRecordingPublisher()
            .sink { [weak self] samples in
                guard let self = self else { return }
                self.asrEngine.acceptWaveform(samples)
                let partial = self.asrEngine.getPendingText()
                if partial.isNotBlank {
                    self.state.partialAsrText = partial
                }
            }
    }

    func stopListening() {
        os_log(.info, "MainVM: stop listening")
        audioRecorder.stop()
        recordingCancellable?.cancel()
        recordingCancellable = nil

        state.voiceState = .recognizing

        recognitionTask?.cancel()
        recognitionTask = Task.detached(priority: .userInitiated) { [weak self] in
            guard let self = self else { return }

            let text = await MainActor.run { self.asrEngine.inputFinished() }

            if text.isEmpty {
                await MainActor.run {
                    os_log(.info, "MainVM: ASR returned blank, returning to idle")
                    self.state.voiceState = .idle
                    self.state.partialAsrText = ""
                }
                return
            }

            await MainActor.run {
                self.state.finalAsrText = text
                self.state.partialAsrText = ""
                self.state.voiceState = .thinking
            }

            // Check config
            let hasConfig = await MainActor.run { self.configRepo.hasConfig }
            if !hasConfig {
                await MainActor.run {
                    self.state.voiceState = .error("请先在设置中配置 API 信息")
                }
                return
            }

            // Stream LLM response
            os_log(.info, "MainVM: sending to LLM, text='%@'", text)
            await self.streamLLMResponse(text)
        }
    }

    func cancelListening() {
        os_log(.info, "MainVM: cancel all active operations")
        audioRecorder.stop()
        recordingCancellable?.cancel()
        recordingCancellable = nil
        recognitionTask?.cancel()
        recognitionTask = nil
        streamingCancellable?.cancel()
        streamingCancellable = nil
        streamingContinuation?.resume()
        streamingContinuation = nil
        speakingTask?.cancel()
        speakingTask = nil
        audioPlayer.stop()
        asrEngine.reset()
        state.voiceState = .idle
        state.partialAsrText = ""
    }

    // MARK: - LLM Streaming

    private func streamLLMResponse(_ text: String) async {
        let streamPublisher = chatSession.sendStream(text)

        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            self.streamingContinuation = continuation
            var fullReply = ""
            var hasCompleted = false

            streamingCancellable = streamPublisher
                .flatMap { $0 }  // unwrap inner publisher
                .sink(
                    receiveCompletion: { [weak self] completion in
                        guard !hasCompleted else { return }
                        hasCompleted = true
                        self?.streamingContinuation = nil

                        Task { @MainActor in
                            if case .failure(let error) = completion {
                                os_log(.error, "MainVM: LLM failed: %{public}@",
                                       error.localizedDescription)
                                self?.state.voiceState = .error(
                                    "请求失败: \(error.localizedDescription)"
                                )
                            } else {
                                self?.chatSession.appendAssistantReply(fullReply)
                                self?.state.assistantReply = fullReply
                                if self?.state.ttsEnabled == true {
                                    self?.speakText(fullReply)
                                } else {
                                    self?.state.voiceState = .idle
                                }
                            }
                        }
                        continuation.resume()
                    },
                    receiveValue: { [weak self] token in
                        fullReply += token
                        Task { @MainActor in
                            self?.state.assistantReply = fullReply
                        }
                    }
                )
        }

        streamingCancellable = nil
        streamingContinuation = nil
    }

    // MARK: - TTS + Playback

    func speakText(_ text: String) {
        guard state.ttsEnabled else {
            state.voiceState = .idle
            return
        }
        speakingTask?.cancel()
        audioPlayer.stop()

        state.voiceState = .speaking

        speakingTask = Task { @MainActor [weak self] in
            guard let self = self else { return }

            let sentences = TextNormalizer.splitSentences(text)
            os_log(.info, "MainVM: %d sentences to speak", sentences.count)
            let sr = self.ttsEngine.sampleRate

            // Synthesise each sentence on a background thread, then
            // concatenate everything into one buffer so we play once.
            // This avoids engine start/stop clicks between sentences.
            let merged: [Float] = await Task.detached(priority: .userInitiated) {
                let silenceSamples = Int(0.3 * Double(sr))  // 300ms gap
                var allSamples: [Float] = []

                for sentence in sentences {
                    if Task.isCancelled { break }
                    let normalized = TextNormalizer.normalize(sentence)
                    guard normalized.isNotBlank else { continue }

                    if let pcm = await self.ttsEngine.synthesize(text: normalized) {
                        allSamples.append(contentsOf: pcm)
                        // Insert silence gap between sentences
                        allSamples.append(contentsOf: Array(repeating: 0, count: silenceSamples))
                    }
                }

                // Fade out last ~10ms to prevent click from abrupt cutoff
                let fadeLen = min(Int(0.01 * Double(sr)), allSamples.count)
                if fadeLen > 0 {
                    for i in 0..<fadeLen {
                        let idx = allSamples.count - fadeLen + i
                        let gain = Float(fadeLen - i) / Float(fadeLen)
                        allSamples[idx] *= gain
                    }
                }

                return allSamples
            }.value

            guard !merged.isEmpty else {
                os_log(.info, "MainVM: nothing to speak (empty TTS output)")
                self.state.voiceState = .idle
                return
            }

            os_log(.info, "MainVM: playing %d samples at %d Hz", merged.count, sr)

            // Single playback of the merged buffer
            await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
                self.audioPlayer.play(pcmFloats: merged, sampleRate: Double(sr)) {
                    cont.resume()
                }
            }

            // Stop the player engine so it doesn't compete with AudioRecorder
            self.audioPlayer.stop()

            self.state.voiceState = .idle
            os_log(.info, "MainVM: speaking complete")
        }
    }

    func stopSpeaking() {
        speakingTask?.cancel()
        audioPlayer.stop()
        state.voiceState = .idle
    }

    // MARK: - Helpers

    func clearError() {
        state.voiceState = .idle
    }

    func clearHistory() {
        chatSession.clear()
        state.assistantReply = ""
        state.partialAsrText = ""
        state.finalAsrText = ""
    }

    deinit {
        audioRecorder.stop()
        audioPlayer.stop()
        engineCleanup?()
        recordingCancellable?.cancel()
        streamingCancellable?.cancel()
        speakingTask?.cancel()
        recognitionTask?.cancel()
    }
}

File: android/app/src/main/java/com/rd/siri/ui/MainViewModel.kt
package com.rd.siri.ui

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.rd.siri.SiriApp
import com.rd.siri.audio.AudioPlayer
import com.rd.siri.audio.AudioRecorder
import com.rd.siri.asr.SherpaAsrEngine
import com.rd.siri.chat.ChatSession
import com.rd.siri.chat.LlmClient
import com.rd.siri.config.ConfigRepository
import com.rd.siri.model.AppState
import com.rd.siri.model.ChatMessage
import com.rd.siri.model.VoiceState
import com.rd.siri.tts.SherpaTtsEngine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class MainViewModel(application: Application) : AndroidViewModel(application) {

    companion object {
        private const val TAG = "SiriApp"
        private val SENTENCE_TERMINATORS = charArrayOf('。', '！', '？', '!', '?', '\n')
    }

    private val audioRecorder = AudioRecorder(application)
    private val audioPlayer = AudioPlayer()
    private val asrEngine = SherpaAsrEngine(application)
    private val ttsEngine = SherpaTtsEngine(application)
    private val configRepository = ConfigRepository(application)
    private val llmClient = LlmClient(configRepository)
    val chatSession = ChatSession(llmClient)

    private val _state = MutableStateFlow(AppState())
    val state: StateFlow<AppState> = _state.asStateFlow()

    private var recordingJob: Job? = null
    private var streamingJob: Job? = null
    private var speakingJob: Job? = null
    private var processingJob: Job? = null

    init {
        Log.i(TAG, "MainViewModel init start")

        // Load TTS preference
        _state.update { it.copy(ttsEnabled = configRepository.isTtsEnabled()) }

        viewModelScope.launch(Dispatchers.IO) {
            _state.update { it.copy(voiceState = VoiceState.Loading("模型加载中…")) }

            // Check native libraries first — if they failed to load, bail out early
            if (!SiriApp.instance.nativeLibrariesLoaded) {
                _state.update {
                    it.copy(voiceState = VoiceState.Error("原生引擎加载失败，请重新安装应用"))
                }
                return@launch
            }

            val asrDeferred = async { asrEngine.initialize() }
            val ttsDeferred = async { ttsEngine.initialize() }

            val asrReady = asrDeferred.await()
            val ttsReady = ttsDeferred.await()
            Log.i(TAG, "ASR engine init done, ready=$asrReady")
            Log.i(TAG, "TTS engine init done, ready=$ttsReady")

            _state.update {
                if (asrReady && ttsReady) {
                    it.copy(voiceState = VoiceState.Idle, enginesReady = true)
                } else {
                    it.copy(voiceState = VoiceState.Error("模型加载失败，请检查模型文件"))
                }
            }
        }
    }

    fun toggleTts(enable: Boolean) {
        configRepository.setTtsEnabled(enable)
        _state.update { it.copy(ttsEnabled = enable) }
        if (!enable) {
            stopSpeaking()
        }
    }

    fun checkConfig(): Boolean {
        val hasConfig = configRepository.hasConfig
        _state.update {
            val newVoiceState = if (hasConfig && it.voiceState is VoiceState.Error
                && it.voiceState.message.contains("配置 API")
            ) {
                VoiceState.Idle
            } else {
                it.voiceState
            }
            it.copy(hasConfig = hasConfig, voiceState = newVoiceState)
        }
        return hasConfig
    }

    /** Cancel every in-flight task so a new recording starts from a clean slate. */
    private fun cancelAllTasks() {
        Log.i(TAG, "cancelAllTasks: aborting all in-progress work")

        // Coroutine jobs
        recordingJob?.cancel()
        recordingJob = null
        streamingJob?.cancel()
        streamingJob = null
        speakingJob?.cancel()
        speakingJob = null
        processingJob?.cancel()
        processingJob = null

        // Audio I/O
        audioRecorder.stopRecording()
        audioPlayer.stop()

        // ASR engine buffer — discard stale samples from cancelled sessions
        asrEngine.reset()

        // ChatSession rollback: if a user message was added but no reply yet,
        // remove the orphaned message so it does not pollute the next conversation.
        val msgs = chatSession.messages.value
        if (msgs.isNotEmpty() && msgs.last().role == ChatMessage.Role.USER) {
            chatSession.clearLastUserMessage()
        }
    }

    fun startListening() {
        if (!asrEngine.isReady) {
            _state.update { it.copy(voiceState = VoiceState.Error("模型未就绪，请先上传模型文件")) }
            return
        }
        if (!AudioRecorder.isPermissionGranted(getApplication())) {
            _state.update { it.copy(voiceState = VoiceState.Error("请授予麦克风权限")) }
            return
        }

        // Cancel everything in flight — ASR, LLM, TTS, etc.
        cancelAllTasks()

        Log.i(TAG, "startListening: begin recording")
        _state.update { it.copy(voiceState = VoiceState.Listening, partialAsrText = "", finalAsrText = "") }

        recordingJob = viewModelScope.launch {
            audioRecorder.startRecording()
                .catch { e ->
                    Log.e(TAG, "startListening: recording error", e)
                    _state.update { it.copy(voiceState = VoiceState.Error("录音错误: ${e.message}")) }
                }
                .collect { samples ->
                    asrEngine.acceptWaveform(samples)
                    val partial = asrEngine.getPendingText()
                    if (partial.isNotBlank()) {
                        Log.d(TAG, "startListening: partial ASR text='$partial'")
                    }
                    _state.update { it.copy(partialAsrText = partial) }
                }
        }
    }

    fun stopListening() {
        Log.i(TAG, "stopListening: stopping recording")
        audioRecorder.stopRecording()
        recordingJob?.cancel()

        _state.update { it.copy(voiceState = VoiceState.Recognizing) }

        processingJob = viewModelScope.launch(Dispatchers.IO) {
            try {
                recordingJob?.join()
                recordingJob = null

                Log.i(TAG, "stopListening: calling inputFinished")
                val text = asrEngine.inputFinished()
                Log.i(TAG, "stopListening: ASR final text='$text'")
                if (text.isBlank()) {
                    Log.w(TAG, "stopListening: ASR returned blank text, returning to idle")
                    _state.update { it.copy(voiceState = VoiceState.Idle, partialAsrText = "") }
                    return@launch
                }

                _state.update { it.copy(finalAsrText = text, partialAsrText = "", voiceState = VoiceState.Thinking) }

                if (!configRepository.hasConfig) {
                    _state.update {
                        it.copy(voiceState = VoiceState.Error("请先在设置中配置 API 信息"))
                    }
                    return@launch
                }

                Log.i(TAG, "stopListening: sending to LLM, text='$text'")
                val result = chatSession.sendStream(text)
                result.onSuccess { flow ->
                    streamingJob = viewModelScope.launch {
                        val fullReply = StringBuilder()
                        try {
                            speakStream(flow, fullReply)
                        } catch (e: kotlinx.coroutines.CancellationException) {
                        } catch (e: Exception) {
                            Log.e(TAG, "stopListening: stream TTS failed", e)
                        }
                        Log.i(TAG, "stopListening: LLM reply complete, len=${fullReply.length}")
                        _state.update { it.copy(voiceState = VoiceState.Idle) }
                    }
                }.onFailure { e ->
                    Log.e(TAG, "stopListening: LLM request failed", e)
                    _state.update { it.copy(voiceState = VoiceState.Error("请求失败: ${e.message}")) }
                }
            } finally {
                processingJob = null
            }
        }
    }

    fun speakText(text: String) {
        if (!_state.value.ttsEnabled) {
            _state.update { it.copy(voiceState = VoiceState.Idle) }
            return
        }
        speakingJob?.cancel()
        audioPlayer.stop()

        _state.update { it.copy(voiceState = VoiceState.Speaking) }

        speakingJob = viewModelScope.launch {
            try {
                val sentences = com.rd.siri.tts.TextNormalizer.splitSentences(text)
                Log.i(TAG, "speakReply: ${sentences.size} sentences, text='${text.take(80)}'")
                val sampleRate = ttsEngine.getSampleRate()
                val channel = Channel<FloatArray>(Channel.BUFFERED)

                coroutineScope {
                    launch(Dispatchers.IO) {
                        for (sentence in sentences) {
                            val normalized = com.rd.siri.tts.TextNormalizer.normalize(sentence)
                            if (normalized.isBlank()) continue
                            Log.i(TAG, "speakReply: synthesizing sentence='${normalized.take(40)}'")
                            val pcm = ttsEngine.synthesize(normalized, sid = 0)
                            if (pcm != null) {
                                channel.send(pcm)
                            }
                        }
                        channel.close()
                    }

                    launch(Dispatchers.IO) {
                        for (pcm in channel) {
                            audioPlayer.play(pcm, sampleRate)
                        }
                    }
                }
            } catch (e: kotlinx.coroutines.CancellationException) {
            } catch (e: Exception) {
                Log.e(TAG, "speakReply: TTS failed", e)
            } finally {
                _state.update { it.copy(voiceState = VoiceState.Idle) }
            }
        }
    }

    private suspend fun speakStream(
        textFlow: Flow<String>,
        accumulator: StringBuilder
    ) {
        val ttsEnabled = _state.value.ttsEnabled

        if (ttsEnabled) {
            _state.update { it.copy(voiceState = VoiceState.Speaking) }
        }

        if (ttsEnabled) {
            val sampleRate = ttsEngine.getSampleRate()
            val sentenceChannel = Channel<String>(Channel.UNLIMITED)
            val pcmChannel = Channel<FloatArray>(Channel.BUFFERED)
            var lastBoundary = 0

            coroutineScope {
                val synthJob = launch(Dispatchers.IO) {
                    try {
                        for (sentence in sentenceChannel) {
                            val normalized = com.rd.siri.tts.TextNormalizer.normalize(sentence)
                            if (normalized.isBlank()) continue
                            Log.i(TAG, "speakStream: synthesizing '${normalized.take(40)}'")
                            val pcm = ttsEngine.synthesize(normalized, sid = 0)
                            if (pcm != null) {
                                pcmChannel.send(pcm)
                            }
                        }
                    } finally {
                        pcmChannel.close()
                    }
                }

                val playerJob = launch(Dispatchers.IO) {
                    for (pcm in pcmChannel) {
                        audioPlayer.play(pcm, sampleRate)
                    }
                }

                try {
                    textFlow.collect { token ->
                        accumulator.append(token)
                        _state.update { it.copy(assistantReply = accumulator.toString()) }

                        val text = accumulator.toString()
                        while (lastBoundary < text.length) {
                            var boundaryIdx = -1
                            for (i in lastBoundary until text.length) {
                                if (text[i] in SENTENCE_TERMINATORS) {
                                    boundaryIdx = i
                                    break
                                }
                            }
                            if (boundaryIdx == -1) break

                            val sentence = text.substring(lastBoundary, boundaryIdx).trim()
                            if (sentence.isNotBlank()) {
                                sentenceChannel.send(sentence)
                            }
                            lastBoundary = boundaryIdx + 1
                        }
                    }
                } finally {
                    val remaining = accumulator.toString().substring(lastBoundary).trim()
                    if (remaining.isNotBlank()) {
                        Log.i(TAG, "speakStream: flushing remaining '${remaining.take(40)}'")
                        sentenceChannel.send(remaining)
                    }
                    sentenceChannel.close()
                }

                val reply = accumulator.toString()
                if (reply.isNotBlank()) {
                    chatSession.appendAssistantReply(reply)
                }
                _state.update { it.copy(assistantReply = reply) }

                synthJob.join()
                playerJob.join()
            }
        } else {
            // TTS disabled: just collect the stream, save to chat, no audio
            try {
                textFlow.collect { token ->
                    accumulator.append(token)
                    _state.update { it.copy(assistantReply = accumulator.toString()) }
                }
            } finally {
                val reply = accumulator.toString()
                if (reply.isNotBlank()) {
                    chatSession.appendAssistantReply(reply)
                }
                _state.update { it.copy(assistantReply = reply) }
            }
        }
    }

    fun stopSpeaking() {
        speakingJob?.cancel()
        streamingJob?.cancel()
        audioPlayer.stop()
        _state.update { it.copy(voiceState = VoiceState.Idle) }
    }

    fun finishSpeaking() {
        Log.i(TAG, "finishSpeaking called")
        stopSpeaking()
    }

    fun cancelListening() {
        Log.i(TAG, "cancelListening called")
        audioRecorder.stopRecording()
        audioPlayer.stop()
        recordingJob?.cancel()
        recordingJob = null
        processingJob?.cancel()
        processingJob = null
        speakingJob?.cancel()
        speakingJob = null
        streamingJob?.cancel()
        streamingJob = null
        asrEngine.reset()
        _state.update { it.copy(voiceState = VoiceState.Idle, partialAsrText = "") }
    }

    fun clearError() {
        _state.update { it.copy(voiceState = VoiceState.Idle) }
    }

    fun clearHistory() {
        chatSession.clear()
        _state.update { it.copy(assistantReply = "", partialAsrText = "", finalAsrText = "") }
    }

    override fun onCleared() {
        super.onCleared()
        audioRecorder.stopRecording()
        audioPlayer.release()
        asrEngine.destroy()
        ttsEngine.destroy()
        recordingJob?.cancel()
        streamingJob?.cancel()
        speakingJob?.cancel()
        processingJob?.cancel()
    }

}
